core/src/muxing.rs

```diff
@@ -231,511 +114,92 @@ pub enum StreamMuxerEvent<T> {
-        substream: &mut Self::OutboundSubstream,
-    ) -> Poll<Result<Self::Substream, Self::Error>> {
-        let mut list = self.outbound.lock();
-        let substream = match self
-            .inner
-            .poll_outbound(cx, list.get_mut(substream).unwrap())
-        {
-            Poll::Pending => return Poll::Pending,
-            Poll::Ready(Ok(s)) => s,
-            Poll::Ready(Err(err)) => return Poll::Ready(Err(err.into())),
-        };
-        let id = self.next_substream.fetch_add(1, Ordering::Relaxed);
-        self.substreams.lock().insert(id, substream);
-        Poll::Ready(Ok(id))
-    }
-
-    #[inline]
-    fn destroy_outbound(&self, substream: Self::OutboundSubstream) {
-        let mut list = self.outbound.lock();
-        self.inner
-            .destroy_outbound(list.remove(&substream).unwrap())
-    }
-
-    #[inline]
-    fn read_substream(
-        &self,
-        cx: &mut Context<'_>,
-        s: &mut Self::Substream,
-        buf: &mut [u8],
-    ) -> Poll<Result<usize, Self::Error>> {
-        let mut list = self.substreams.lock();
-        self.inner
-            .read_substream(cx, list.get_mut(s).unwrap(), buf)
-            .map_err(|e| e.into())
-    }
-
-    #[inline]
-    fn write_substream(
-        &self,
-        cx: &mut Context<'_>,
-        s: &mut Self::Substream,
-        buf: &[u8],
-    ) -> Poll<Result<usize, Self::Error>> {
-        let mut list = self.substreams.lock();
-        self.inner
-            .write_substream(cx, list.get_mut(s).unwrap(), buf)
-            .map_err(|e| e.into())
-    }
-
-    #[inline]
-    fn flush_substream(
-        &self,
-        cx: &mut Context<'_>,
-        s: &mut Self::Substream,
-    ) -> Poll<Result<(), Self::Error>> {
-        let mut list = self.substreams.lock();
-        self.inner
-            .flush_substream(cx, list.get_mut(s).unwrap())
-            .map_err(|e| e.into())
-    }
-
-    #[inline]
-    fn shutdown_substream(
-        &self,
-        cx: &mut Context<'_>,
-        s: &mut Self::Substream,
-    ) -> Poll<Result<(), Self::Error>> {
-        let mut list = self.substreams.lock();
-        self.inner
-            .shutdown_substream(cx, list.get_mut(s).unwrap())
-            .map_err(|e| e.into())
-    }
+    /// Allow this state machine to make progress.
+    ///
+    /// In case a new substream was requested, we invoke the provided poll function and return the
+    /// created substream along with its ID.
+    pub fn poll<S, E>(
+        &mut self,
+        poll_open: impl FnOnce() -> Poll<Result<S, E>>,
+    ) -> Poll<Result<(S, OutboundSubstreamId), E>> {
+        if self.substreams_to_open.is_empty() {
+            return Poll::Pending;
+        }
 
-    #[inline]
-    fn destroy_substream(&self, substream: Self::Substream) {
-        let mut list = self.substreams.lock();
-        self.inner
-            .destroy_substream(list.remove(&substream).unwrap())
-    }
+        let stream = futures::ready!(poll_open())?;
```

Comment:
This indirection seems counterintuitive to me. While more code, I would expect the muxer calling the `OutboundStreams` instead of the `OutboundStreams` calling the muxer to be easier to follow.

Reply:
The _concrete_ muxers are calling the `OutboundSubstreams` component.

mplex: https://github.com/thomaseizinger/rust-libp2p/blob/db7cb6b47ace2f844a7a6da6ece345d8d2f6657a/muxers/mplex/src/lib.rs#L118-L120
yamux: https://github.com/thomaseizinger/rust-libp2p/blob/db7cb6b47ace2f844a7a6da6ece345d8d2f6657a/muxers/yamux/src/lib.rs#L110-L113

Reply:
I am not sure how to otherwise design this abstraction so it can be reused by each muxer implementation.

Reply:
I am more referring to the closer (`poll_open`) passed by the muxer to the `OutboundSubstreams`. Is that a strict requirement?

Reply:
I am not sure I completely follow but here is the thinking:

1. We should only poll to open a new open substream if the user requested one.
2. Requesting to open a new substream issues an `OutboundSubstreamId`.
3. Thus, opening a new substream is based on the state of having issued `OutboundSubstreamId`s
4. The correct IDs need to be mapped to the `OutboundSubstreamId`s that were handed out.

I could inline this abstraction but it felt a bit error prone to duplicate the behaviour of only polling for new substreams if we have issued IDs but not serviced them yet.